examples/token/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+#[inline]
+pub fn symbol(state: State) -> RawBytes {
+    match String::from_utf8(state.symbol) {
+        Ok(s) => encode!(s.into_bytes()),
+        Err(err) => abort!(ErrIllegalState, "could not parse name as UTF-8: {:?}", err),
+    }
+}
+
+/// Returns the maximum supply. Method number 3.
+#[inline]
+pub fn max_supply(state: State) -> RawBytes {
+    encode!(BigIntSer(&state.max_supply))
+}
+
+/// The input parameters for a transfer.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct TransferParams {
+    pub recipient: Address,
+    #[serde(with = "bigint_ser")]
+    pub amount: TokenAmount,
+}
+
+impl Cbor for TransferParams {}
+
+/// Transfer a token amount.
+pub fn transfer(mut state: State, params: TransferParams) {
+    // Load the balances HAMT.
+    // TODO Using BitIntDe because it's both Ser and De; this is a misnomer and
+    //  we should fix it.
+    let mut balances =
+        match Hamt::<Blockstore, BigIntDe, ActorID>::load(&state.balances, Blockstore) {
+            Ok(map) => map,
+            Err(err) => abort!(ErrIllegalState, "failed to load balances hamt: {:?}", err),
+        };
+
+    // Load the sender's balance.
+    let sender_id = fvm_sdk::message::caller();
+    let mut sender_bal = match balances.get(&sender_id) {
+        Ok(Some(bal)) => bal.clone(),
+        Ok(None) => BigIntDe(TokenAmount::zero()),
+        Err(err) => abort!(ErrIllegalState, "failed to get balance: {:?}", err),
+    };
+
+    // Sender has insufficient balance.
+    if sender_bal.0 < params.amount {
+        abort!(ErrInsufficientFunds, "sender has insufficient balance")
+    }
+
+    // Resolve the recipient into an ID address.
+    // TODO See addressing section on module docs.
+    let recipient_id = match fvm_sdk::actor::resolve_address(&params.recipient) {
+        Some(id) => id,
+        None => abort!(ErrIllegalArgument, "failed to resolve address"),
+    };
+
+    // Load the recipient's balance.
+    let mut recipient_bal = match balances.get(&recipient_id) {
+        Ok(Some(bal)) => bal.clone(),
+        Ok(None) => BigIntDe(TokenAmount::zero()),
+        Err(err) => abort!(
+            ErrIllegalState,
+            "failed to query hamt when getting recipient balance: {:?}",
+            err
+        ),
+    };
+
+    // Update balances.
+    sender_bal.0 -= &params.amount;
+    recipient_bal.0 += &params.amount;
+
+    // Set the updated sender balance in the balances HAMT.
+    if let Err(err) = balances.set(sender_id, sender_bal.clone()) {
+        abort!(
+            ErrIllegalState,
+            "failed to set new sender balance in balances hamt: {:?}",
+            err
+        )
+    }
+
+    // Set the updated recipient balance in the balances HAMT.
+    if let Err(err) = balances.set(recipient_id, recipient_bal.clone()) {
+        abort!(
+            ErrIllegalState,
+            "failed to set new recipient balance in balances hamt: {:?}",
+            err
+        )
+    }
```

Comment:
This will print tokens from thin air if the recipient is the sender. Writing the recipient balance overwrites the reduced sender balance.

An example of a safe sequence is https://github.com/filecoin-project/specs-actors/blob/master/support/vm/vm.go#L546

Reply:
Well spotted. Fixed in [`59cc833` (#290)](https://github.com/filecoin-project/ref-fvm/pull/290/commits/59cc8338c2c3ff504a978047fc9d0f0b51e05903), along with other logic changes.

The goal of this exercise is not to achieve complete safety, but rather to illustrate the current state of deploying custom actors for local testing. But it's good to know that we have a safe sequence others can use as a reference.